hack/golangci-hints.yaml

```diff
@@ -107,6 +107,12 @@ linters-settings: # please keep this alphabetized
     - p: \.Extract
       pkg: ^k8s\.io/client-go/applyconfigurations/
       msg: should not be used because managedFields was removed
+    - p: \.HandleError
+      pkg: ^k8s\.io/apimachinery/pkg/util/runtime
+      msg: should consider HandleErrorInContext in new or modified code
+    - p: \.HandleCrash
+      pkg: ^k8s\.io/apimachinery/pkg/util/runtime
+      msg: should consider HandleCrashInContext in new or modified code
```

Comment:
I was reluctant to declare `HandleError` and `HandlePanic` as deprecated. That leads to confusion and unnecessary work in downstream projects which are fine with the existing functions.

Instead, hints that only show up in `pull-kubernetes-linter-hints` get defined here.

Reply:
We have the same problem whenever we introduce a new API which should be used instead of some old, still supported one. Even if we were to deprecate the old one, we still wouldn't know exactly where in the existing code base we should replace it.

Let me propose something else:
- Instead of `// Deprecated: Use HandleCrashWithContext` for `HandleCrash`, we add `//logcheck:context // Use HandleCrashWithContext instead in code which supports contextual logging.`.
- The logcheck linter gets extended to recognize this special comment for functions and methods.
- If, and only if, some code is expected to be "contextual" (defined in `hack/logcheck.conf`) and uses such an API, then logcheck will report it.

This is conceptually similar to how it warns about some known klog API calls that shouldn't be used anymore, except that the knowledge about those calls is not baked into the linter.

@sttts how does this sound?

Reply:
There is one difference compared to the hints: as soon as we mark an API like this, we have to fix code which currently is supposed to be contextual (kube-controller-manager, kube-scheduler, etc.). I think that's doable.

Reply:
Implemented in https://github.com/kubernetes-sigs/logtools/pull/25.

Reply:
no strong opinion. But yes, sounds reasonable.

Reply:
I'm doing the conversion for kube-controller now to verify that it's doable. One observation: there are a lot of `HandleError` calls which should never be reached (checked type conversions, getting key for object, accessing local cache). I'm not going to convert those because introducing a context in the call chain just for dead code isn't worth the effort and overhead. Instead, I'll add `nolint:logcheck` with comment to indicate that the call has been checked and is okay *in those places*.